### PR TITLE
Update fs-extra to v8

### LINKED ISSUE #64: Random build failures due to fs-extra defect
# Bug Report

### What is expected to happen?
You are able to build your Cordova project reliably.

### What does actually happen?
Builds randomly fail with the error **Source and destination must not be the same.**

## Information
The random failures are caused by a [defect in fs-extra](https://github.com/jprichardson/node-fs-extra/issues/657), which neglects to respect an inode-related issue in Node itself.

### Command or Code
Every command that copies files with `fs-extra` is affected. `cordova prepare` is most likely to cause this.

### Environment, Platform, Device
Windows 10

### Version information
Cordova 8.1.2 (cordova-lib@8.1.1)

## Checklist

- [X] I searched for existing GitHub issues
- [X] I updated all Cordova tooling to most recent version
- [X] I included all the necessary information above

### PR DESCRIPTION
**Pull request by dpogue**
### Platforms affected
All, but hopefully mostly Windows

### Motivation and Context

Hopefully resolves ~~#64~~ _apache/cordova#121 on cordova-common only_

### Description

Updated `fs-extra` to the latest version (8.0.0).

### Testing

`npm test` passed locally on Linux, will ensure it's green on CI.

### Checklist

- [x] I've run the tests to see all new and existing tests pass
- [x] I added automated test coverage as appropriate for this change
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))

### CODE REVIEW

**Comment by raphinesse** [comment]
AFAICT, just updating will not fix anything. I think the new version just includes a workaround to disable some checks during `copy*`. The downside being that this could delete your copy source. And that we'd have to edit each and every location where we copy something.

Or is that information outdated?

**Comment by FlorinSerban** [comment]
> AFAICT, just updating will not fix anything. I think the new version just includes an opt-in workaround to disable some checks during `copy*`. The downside being that this could delete your copy source. And that we'd have to edit each and every location where we copy something.
> 
> Or is that information outdated?

I think the right fix was implemented. Please check this [comment](https://github.com/jprichardson/node-fs-extra/pull/667#issuecomment-490404722).

**Comment by raphinesse** [comment]
Looks like you're right, @FlorinSerban. Thanks for the pointer!

**Comment by catalinmitrofan** [comment]
When are you guys planning a release with this fix?

**Comment by oliversalzburg** [comment]
@catalinmitrofan It hasn't even been merged yet. Calm down 😁

**Comment by catalinmitrofan** [comment]
@oliversalzburg, I just want to know, when the changes will be merged => when is a new release planned?

**Comment by oliversalzburg** [comment]
I believe there is no schedule. If this is a blocking issue for you, I recommend [implementing a workaround](https://github.com/apache/cordova-common/issues/64#issuecomment-490427299) for the time being.

**Comment by dimitriscsd** [comment]
Considering the severity of the issue (Everyone with a windows build server is unable to build their cordova project), I believe there should be a schedule asap.

The workaround is a nice band aid but a proper fix is always better.

**Comment by oliversalzburg** [comment]
@dimitriscsd We have waited a long time for fs-extra to fix their code. There is really no need for this kind of pressure so soon after a change has been proposed in Cordova. I'm sure everyone agrees that this needs to be fixed ASAP.

**Comment by dimitriscsd** [comment]
That's perfect. I didn't mean to apply pressure, I just pointed out that it's a serious problem users of Cordova are facing and it should be places somewhere in the pipeline rather than be left in a backlog somewhere.

It's great to hear that you guys understand the severity of the issue and I'm grateful for your work and efforts.

**Comment by brodybits** [comment]
Let's get this one merged and published. Thanks guys for the patience.

**Comment by dpogue** [comment]
Note that we also need to update fs-extra in any other modules that use it, which includes cordova-lib, cordova-create, and cordova-fetch, and possibly others

**Comment by catalinmitrofan** [comment]
Thanks again to everyone for their fast work on this issue.

**Comment by catalinmitrofan** [comment]
So, is there a planned release with this fix?

**Comment by brodybits** [comment]
> So, is there a planned release with this fix?

I suggest you post this kind of question to the forum at: <dev@cordova.apache.org>

I would like to clarify a couple things:

To fully "release" this fix, we would actually need to apply this fix and make a release for multiple components as @dpogue said in <https://github.com/apache/cordova-common/pull/70#issuecomment-491838840>.

I personally do not have so much time to make a new release of the components right now. Keep in mind that this is a 100% volunteer project with very limited corporate sponsorship, if any. (I think a few big companies had donated maintainer time in the past, seems to be gone now.)

Thanks for your continued patience.

**Comment by catalinmitrofan** [comment]
@brodybits , thank you for the explanation and your answer. I will wait then for the next release

**Comment by l3ender** [comment]
Thanks for all those who have worked on this. Is there anything I could do to help out to get it merged and released?

File: package.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     "cross-spawn": "^6.0.5",
     "elementtree": "0.1.7",
     "endent": "^1.1.1",
-    "fs-extra": "^7.0.0",
+    "fs-extra": "^8.0.0",
```
Comment (brodybits) [inline package.json]:
```suggestion
    "fs-extra": "^8.0.1",
```

Any objections if we apply this change, now that fs-extra@8.0.1 is out?

Comment (GedasGa) [reply package.json]: Not necessary. Using the caret will update you to the most recent major version. `^8.0.0` will match any 8.x.x release including 8.0.1

Comment (brodybits) [reply package.json]: @GedasGa that is correct. My idea is that the spec should be updated if possible to be consistent with the other Cordova packages where we should apply the latest fs-extra version spec. Not a blocker on my part, though.

**Comment by brodybits** [comment]
I would like to merge this, maybe with my fs-extra@^8.0.1 suggestion if there are no objections from other maintainers. Then one of us maintainers will have to publish a new cordova-common release in the next 1-2 weeks.

@l3ender and others please followup with us via email to <dev@cordova.apache.org> in case we drop this one. I think all maintainers are extremely busy in general.

**Comment by dpogue** [comment]
As I understand it, the reason this has not been merged is because nobody has been able to actually confirm whether it fixes the problem or not (although it's very hard to test that when other libraries are also pulling in their own copies of fs-extra).

As long as we're confident that the major version bump of fs-extra doesn't break anything here, I'm good with seeing this merged.

**Comment by brodybits** [comment]
I would *personally* favor publishing this change as a *minor* release; fs-extra@8 seems to not break for Node.js back to version 6. Any other comments or objections?

**Comment by catalinmitrofan** [comment]
@dpogue , I have moved the cordova-common repository to our own, and I patched it there. Since I patched it, I didn't get any issues when trying to build the app for windows. I have been using it like this for the last  20 days, without any issues. When the change will be published, then I will revert back to using the repository from here, until then, I am using my "patched" repository.

**Comment by brodybits** [comment]
I have just merged this proposal, with an updated title since it will include latest patch and minor release on fs-extra@8. I will now propose a minor release to be published within the next week.

**Comment by brodybits** [comment]
This update is now published in cordova-common@3.2.0. I documented the recommended update procedure in <https://github.com/apache/cordova/issues/121#issuecomment-502143314>.

**Comment by gwynjudd** [comment]
I'm still getting this issue with the updated version of cordova-common and fs-extra. Things I've checked:

1. version of cordova-common is 3.2.0
2. version of fs-extra is 8.1.0

node version is 8.9.1

Any other things I should be checking?

**Comment by brodybits** [comment]
It is more likely to work on Node.js 10 which supports bigint. There may be some more workarounds for older Node.js versions but I would consider it to be a waste of time.

**Comment by gwynjudd** [comment]
@brodybits thanks I'll give it a shot

**Pull request merged**